### PR TITLE
initial commit for scheduled anonymizer

### PR DESCRIPTION
**Pull request by ymao2**
## Description

The basic idea for making the job of anonymising the live data and the job of restoring anonymised db into qa is 
 - Run a scheduled job on live env to get anonymised db into QA bucket on weekend by connection to replication DB
   the job of anonymising the whole database will be split into small tasks per table , if the size of the data is big, then the task of anonymising a table will be split further, all those tasks will be run in parallel via sidekiq to shorter the process
 - Run a scheduled job a few hours later to download the anonymised data from QA bucket and restore them into QA DB at the same of the first job
 - Run a scheduled job to update the search index every half an hour at the same day
plus the following change 
- Add the part for anonymising the Contact table
The further task for this ticket is to create a secret file in live prod for storing the credentials for target bucket in QA env. this will be done manually and outside this PR as the secret file is not stored with repo anymore
## Self-review checklist

* [ ] (1) Quick stakeholder demo done OR
* [ ] (2) ...bug with before and after screenshots
* [X] (3) Tests passing
* [X] (4) Branch ready to be merged (not work in progress)
* [X] (5) No superfluous changes in diff
* [X] (6) No TODO's without new ticket numbers
* [X] (7) PR Prefixed with ticket number e.g. `CT-7654 ...`
* [X] (8) Data migration script is created if any of letter templates is changed

### Related JIRA tickets

https://dsdmoj.atlassian.net/jira/software/c/projects/CT/boards/25?modal=detail&selectedIssue=CT-3728

### CODE REVIEW

File: lib/db/database_dumper.rb [hunk 1]
```diff
@@ -2,168 +2,146 @@
...
-  def initialize(anonymize, tag, is_store_to_s3_bucket, s3_bucket: nil)
-    @anonymize = anonymize
-    @anonymizer = nil
-    @db_connection_url = ENV['DATABASE_URL'] || 'postgres://localhost/correspondence_platform_development'
-    @s3_bucket = s3_bucket
-    @outcome_files = []
+
+  def initialize(tag, running_mode = 'tasks', is_store_to_s3_bucket = true, s3_bucket_setting = nil)
+    ####
+    # tag:  part of the file name for  the anonymised data, it is ID of the anonymised db copy on s3 bucket
```
Comment (willmcb) [inline lib/db/database_dumper.rb:19]: We should probably add this to a "how to" document of some sort.

**Review by willmcb — APPROVED** [review]
LGTM 👍 few comments we can discuss in out catch up

File: lib/db/database_anonymizer_tasks.rb [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+  def task_dump_pre_data_tables(_)
+    filename = "#{@base_file_name}_00_pre_data.sql"
+    command_line = "pg_dump #{@db_connection_url} -v --no-owner --no-privileges --no-password -O --section=pre-data -f #{filename}"
+    result = system command_line
+    raise 'Unable to execute pg_dump command' unless result == true
+    filename 
+  end
+
+  def task_dump_post_data_tables(_)
```
[the remark this answers is not in the document]
Comment (ymao2) [reply lib/db/database_anonymizer_tasks.rb]: :(, forgot this task, added it

File: lib/db/database_dumper.rb (hunk 1 above)
Comment (ymao2) [reply lib/db/database_dumper.rb:19]: Will add a follow-up ticket

**Pull request merged**